File: app/src/main/java/com/karla/learningverbs/utils/base/BaseActivity.java
package com.karla.learningverbs.utils.base;

import android.app.Dialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.viewbinding.ViewBinding;

import com.karla.learningverbs.R;

public abstract class BaseActivity<BINDING extends ViewBinding, VM extends BaseViewModel> extends AppCompatActivity {

    protected VM viewModel;
    protected BINDING binding;
    private Dialog dialog;

    protected abstract VM createViewModel();

    @NonNull
    protected abstract BINDING createViewBinding(LayoutInflater layoutInflater);

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = createViewBinding(LayoutInflater.from(this));
        setContentView(binding.getRoot());
        dialog = new Dialog(BaseActivity.this, R.style.customLottie);
        viewModel = createViewModel();
        observers();
    }


    private void observers() {
        viewModel.loading.observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean isLoading) {
                if (isLoading) {
                    showProgress();
                } else {
                    hideProgress();
                }
            }
        });

        viewModel.msgError.observe(this, new Observer<String>() {
            @Override
            public void onChanged(String msgError) {
                showMessageError(msgError);
            }
        });
    }

    void showProgress() {
        dialog.setContentView(R.layout.loading_view);
        if (!dialog.isShowing()) dialog.show();
    }

    void hideProgress() {
        if (dialog.isShowing()) dialog.dismiss();
    }


    public void showMessageError(String msgError) {
        Toast.makeText(BaseActivity.this, msgError, Toast.LENGTH_SHORT).show();
    }

}

File: app/src/main/java/com/karla/learningverbs/kotlin/utils/base/BaseActivity.kt
package com.karla.learningverbs.kotlin.utils.base

import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.viewbinding.ViewBinding
import com.karla.learningverbs.R

abstract class BaseActivity<BINDING : ViewBinding, VM : BaseViewModel> :AppCompatActivity() {

    protected lateinit var viewModel: VM
    protected lateinit var binding: BINDING
    private lateinit var dialog: Dialog


    protected abstract fun createViewModel(): VM

    protected abstract fun createViewBinding(layoutInflater: LayoutInflater?): BINDING

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = createViewBinding(LayoutInflater.from(this))
        setContentView(binding.root)
        viewModel = createViewModel()
        dialog = Dialog(this@BaseActivity, R.style.customLottie)
        observers()
    }

    private fun observers() {
        viewModel.loading.observe(this, object : Observer<Boolean?> {
            override fun onChanged(isLoading: Boolean?) {
                if (isLoading!!) {
                    showProgress()
                } else {
                    hideProgress()
                }            }
        })
        viewModel.msgError.observe(this, object : Observer<String?> {
            override fun onChanged(msgError: String?) {
                showMessageError(msgError!!)
            }
        })
    }

    open fun showProgress() {
        dialog!!.setContentView(R.layout.loading_view)
        if (!dialog!!.isShowing) dialog!!.show()
    }

    open fun hideProgress() {
        if (dialog!!.isShowing) dialog!!.dismiss()
    }


    open fun showMessageError(msgError: String?) {
        Toast.makeText(this@BaseActivity, msgError, Toast.LENGTH_SHORT).show()
    }

}